ob-deno.el

```diff
@@ -97,22 +97,54 @@ which can be specified by VALUES."
     (format "--allow-%s" allow-param)))
 
 (defun ob-deno-read (results)
-  "Convert RESULTS into an appropriate elisp value.
-If RESULTS look like a table, then convert them into an
-Emacs-lisp table, otherwise return the results as a string."
-  (org-babel-read
-   (if (and (stringp results)
-	    (string-prefix-p "[" results)
-	    (string-suffix-p "]" results))
-       (org-babel-read
-        (concat "'"
-                (replace-regexp-in-string
-                 "\\[" "(" (replace-regexp-in-string
-                            "\\]" ")" (replace-regexp-in-string
-                                       ",[[:space:]]" " "
-				       (replace-regexp-in-string
-					"'" "\"" results))))))
-     results)))
+  "Convert RESULTS into an Org-mode table recognized by Org Babel.
+Handles JSON arrays of objects, arrays, primitives, and single objects.
+Preserves key order and empty rows."
+  (let ((json-array (json-parse-string results)))
+    (cond
+     ;; Case 1: Single Object (hash table)
+     ((hash-table-p json-array)
+      (let ((keys '())
+            (row '()))
+        ;; Collect keys and build row in insertion order
+        (maphash (lambda (key value)
+                   (setq keys (append keys (list key)))
+                   (setq row (append row (list (if value (format "%s" value) "")))))
+                 json-array)
+        (append (list keys 'hline) (list row))))
+
+     ;; Case 2: Array of Objects (vector of hash tables)
+     ((and (vectorp json-array) (hash-table-p (aref json-array 0)))
+      (let ((keys '()))
+        ;; Collect all keys in order of appearance across all objects
+        (mapc (lambda (obj)
+                (maphash (lambda (key _)
+                           (unless (member key keys)
+                             (setq keys (append keys (list key)))))
+                         obj))
+              json-array)
+        (let ((rows (mapcar (lambda (obj)
+                              (mapcar (lambda (key)
+                                        (let ((value (gethash key obj)))
+                                          (if value (format "%s" value) "")))
+                                      keys))
+                            (append json-array nil))))
+          (append (list keys 'hline) rows))))
+
+     ;; Case 3: Array of Arrays (vector of vectors)
+     ((and (vectorp json-array) (vectorp (aref json-array 0)))
+      (mapcar (lambda (row)
+                (if (seq-empty-p row)
+                    (make-list (length (aref json-array 0)) "")
+                  (append row nil)))
+              json-array))
+
+     ;; Case 4: Array of Primitives (flat vector)
+     ((vectorp json-array)
```

Comment:
Not as nice as others but maybe this way we can ensure we only have primitive values in the array.

```suggestion
     ((and (vectorp json-array) (seq-every-p (lambda (x) (not (or (arrayp x) (hash-table-p x)))) json-array))
```

Reply:
I guess for me its quite ok / possibly even preferred if the heterogenous array is undefined behavior and you see it try and fail to format

Reply:
I understand the reasoning but I believe it is not the responsibility of ob-deno. Also I believe seeing the JSON output is much helpful on those cases. Otherwise you *have to* guess the issue is arising from an heterogeneous array (which is hard to guess because they don't know about the implementation).

Another example, if you were to use an array like a tuple to return multiple different values (which is a valid and common use case) then you need to change your code to make it compatible with ob-deno needlessly. Of course even with these changes it is still possible to have formatting errors/issues but I believe it is better to minimize them so that user is not interrupted and not required to fiddle with ob-deno configuration as much as possible.